### PR TITLE
xLearn for Windows

### PR DESCRIPTION
**Pull request by etveritas**
@aksnzhy Hello, I have modified the format to make it easy to review where have changed about the code, and add doc about installation for windows.

### CODE REVIEW

**Comment by aksnzhy** [comment]
That's great!

**Comment by aksnzhy** [comment]
@etveritas Hi，I‘m a little busy these days and I will  review this PR as soon as possible. After the Windows version released, I will send you an invitation and I hope you can be a committer of xLearn, the goal of which is to gain the most impact for large-scale sparse data processing!

**Comment by etveritas** [comment]
@aksnzhy okay

**Comment by etveritas** [comment]
About this latest commit:
1. simplify console setting on windows.
2. only set console text color when console is default setting.

File: CMakeLists.txt [hunk 1]
```diff
@@ -30,8 +30,12 @@ cmake_minimum_required(VERSION 2.8)
 # Using c++11;
 # Optimization on architecture
 #-------------------------------------------------------------------------------
+if(NOT WIN32)
 add_definitions("-Wall -Wno-sign-compare -O3 -std=c++11 
 -march=native -Wno-strict-aliasing -Wno-comment")
+else(WIN32)
+add_definitions("/WX- /MT")
```
Comment (aksnzhy) [inline CMakeLists.txt]: Does this definitions use the best compiler optimization for speed? Sorry, I 'm not a Windows developer.

File: build.bat [hunk 2]
```diff
@@ -0,0 +1,6 @@
+@echo off
+mkdir build && cd build && ^
+cmake -G "Visual Studio 15 Win64" ../ && "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x64 && ^
+MSBuild xLearn.sln /p:Configuration=Release && ^
```
Comment (aksnzhy) [inline build.bat]: What happen if a user doesn't install his visual-studio in C: ?

File: src/base/utsname.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		strcpy(name->machine, "x86");
+		break;
+	case PROCESSOR_ARCHITECTURE_UNKNOWN:
+	default:
+		strcpy(name->machine, "unknown");
+	}
+
+	return 0;
+}
+
```
Comment (aksnzhy) [inline src/base/utsname.h]: Please change the four spaces to two.

**Comment by aksnzhy** [comment]
> @aksnzhy Hello, I have modified the format to make it easy to review where have changed about the code, and add doc about installation for windows.

Thanks for the hard work! I just give some comments for the code-review. Let's improve and polish this PR until merge it to 0.42 version. This will be a major milestone for this project.

File: CMakeLists.txt (hunk 1 above)
Comment (etveritas) [reply CMakeLists.txt]: No，because gtest version is old, add  ``/WX-``  to avoid warning error.

File: build.bat (hunk 2 above)
Comment (etveritas) [reply build.bat]: emm, so, we should build .whl for windows, and users can install xLearn by pip

File: src/base/utsname.h (hunk 3 above)
Comment (etveritas) [reply src/base/utsname.h]: Modified it.

**Comment by aksnzhy** [comment]
@etveritas Hi, does this PR ready for merge?

File: CMakeLists.txt (hunk 1 above)
Comment (etveritas) [reply CMakeLists.txt]: There is no option about ``/O3`` in VS, and it recommend ``/O2`` for release build.

File: build.bat (hunk 2 above)
Comment (etveritas) [reply build.bat]: I build binary with AppVeyor. This is the file url https://ci.appveyor.com/project/etveritas/xlearn/branch/etveritas-xlearn/artifacts.
we can add xLearn to  travis or  AppVeyor for Continuous Integration.

**Comment by etveritas** [comment]
@aksnzhy Maybe I should add some .bat file for ``demo``

**Comment by aksnzhy** [comment]
@etveritas Okay, no problem.

**Comment by etveritas** [comment]
@aksnzhy Finish it.

**Comment by aksnzhy** [comment]
@etveritas Nice! I merged this PR to the master branch. I will release a new version today. Thanks for your hard work!

**Comment by etveritas** [comment]
@aksnzhy Great!

**Comment by aksnzhy** [comment]
@etveritas The 0.4.2 has been released, and I send an invitation to you for inviting you become the xLearn committer. Please check your email. Thanks again for this work!

**Comment by etveritas** [comment]
@aksnzhy accepted

**Comment by jcr-lyxor** [comment]
Just installed on windows with your [guide](https://xlearn-doc.readthedocs.io/en/latest/install/install_windows.html) everything works perfectly great job.

**Pull request merged**